crates/bevy_text/src/text.rs

```diff
@@ -23,160 +24,152 @@ impl Default for CosmicBuffer {
+    ///
+    /// This is private because buffer contents are always refreshed from ECS state when writing glyphs to
+    /// `TextLayoutInfo`. If you want to control the buffer contents manually or use the `cosmic-text`
+    /// editor, then you need to not use `TextBlock` and instead manually implement the conversion to
+    /// `TextLayoutInfo`.
+    pub(crate) buffer: CosmicBuffer,
+    /// Entities for all text spans in the block, including the root-level text.
+    ///
+    /// The [`TextEntity::depth`] field can be used to reconstruct the hierarchy.
+    pub(crate) entities: SmallVec<[TextEntity; 1]>,
+    /// Flag set when any change has been made to this block that should cause it to be rerendered.
+    ///
+    /// Includes:
+    /// - [`TextBlock`] changes.
+    /// - [`TextStyle`] or `Text2d`/`Text`/`TextSpan2d`/`TextSpan` changes anywhere in the block's entity hierarchy.
+    // TODO: This encompasses both structural changes like font size or justification and non-structural
+    // changes like text color and font smoothing. This field currently causes UI to 'remeasure' text, even if
+    // the actual changes are non-structural and can be handled by only rerendering and not remeasuring. A full
+    // solution would probably require splitting TextBlock and TextStyle into structural/non-structural
+    // components for more granular change detection. A cost/benefit analysis is needed.
+    pub(crate) needs_rerender: bool,
+}
+
+impl ComputedTextBlock {
+    /// Accesses entities in this block.
+    ///
+    /// Can be used to look up [`TextStyle`] components for glyphs in [`TextLayoutInfo`] using the `span_index`
+    /// stored there.
+    pub fn entities(&self) -> &[TextEntity] {
+        &self.entities
+    }
+
+    /// Indicates if the text needs to be refreshed in [`TextLayoutInfo`].
+    ///
+    /// Updated automatically by [`detect_text_needs_rerender`](crate::detect_text_needs_rerender) and cleared
+    /// by [`TextPipeline`] methods.
+    pub fn needs_rerender(&self) -> bool {
+        self.needs_rerender
+    }
+}
+
+impl Default for ComputedTextBlock {
+    fn default() -> Self {
+        Self {
+            buffer: CosmicBuffer::default(),
+            entities: SmallVec::default(),
+            needs_rerender: true,
+        }
+    }
+}
+
+/// Component with text format settings for a block of text.
+///
+/// A block of text is composed of text spans, which each have a separate string value and [`TextStyle`]. Text
+/// spans associated with a text block are collected into [`ComputedTextBlock`] for layout, and then inserted
+/// to [`TextLayoutInfo`] for rendering.
+///
+/// See [`Text2d`] for the core component of 2d text, and `Text` in `bevy_ui` for UI text.
 #[derive(Component, Debug, Clone, Default, Reflect)]
 #[reflect(Component, Default, Debug)]
-pub struct Text {
-    /// The text's sections
-    pub sections: Vec<TextSection>,
+#[require(ComputedTextBlock, TextLayoutInfo)]
+pub struct TextBlock {
```

Comment:
The name TextBlock doesn't seem very intuitive, the name implies to me it would be the primary text component holding the string, not the arrangment settings?

Reply:
It's not great but this is what we have in the [rework discussion](https://github.com/bevyengine/bevy/discussions/15014) so I will leave it as-is unless something better is proposed.

Reply:
`TextBlock` is pretty awkward in practice, from my experience migrating examples. `TextLayout` would make a lot more sense. You don't actually add `TextBlock` to an entity and get a text block... you get a partial text block that needs a root text component and text style to function.

Reply:
> you get a partial text block that needs a root text component and text style to function.

Not super important, but when I pitched this design I intended for TextBlock to be the "source of truth text span receiver component" that served its function independently of specific "root text" components. Of course I didn't explicitly call that out and I think the current impl is totally fine. But if we went that route, it would make TextBlock make sense (and would also probably make building future TextBlock types easier). Plenty of open questions there though in terms of implementation / I don't think we need to go down that road in this PR.